### PR TITLE
Storage Statistics

### PR DESCRIPTION
**Pull request by Nivedhita-Chandran**
feat: Implemented Storage Statistics in GoogleHadoopFileSystem

### CODE REVIEW

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/AbstractGhfsStatisticsSource.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+  public IOStatisticsStore getIOStatistics() {
+    return ioStatistics;
+  }
+
+  /**
+   * Setter. this must be called in the subclass constructor with whatever
+   *
+   * @param statistics statistics to set
+   */
+  protected void setIOStatistics(final IOStatisticsStore statistics) {
```
Comment (medb) [inline gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/AbstractGhfsStatisticsSource.java:45]: We usually do not use `final` modifiers for method parameters.

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/AbstractGhfsStatisticsSource.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+  public long incGauge(String name) {
+    return incGauge(name, 1);
+  }
+
+  @Override
+  public String toString() {
+    final StringBuilder sb = new StringBuilder(this.getClass().getSimpleName() + "{");
+    sb.append(ioStatistics);
+    sb.append('}');
+    return sb.toString();
```
Comment (medb) [inline gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/AbstractGhfsStatisticsSource.java:93]: Consider using https://guava.dev/releases/31.0.1-jre/api/docs/com/google/common/base/MoreObjects.ToStringHelper.html

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GhfsInstrumentation.java [hunk 3]
```diff
@@ -0,0 +1,770 @@
...
+    }
+    if (!(metric instanceof MutableCounterLong)) {
+      throw new IllegalStateException(
+          "Metric "
+              + name
+              + " is not a MutableCounterLong: "
+              + metric
+              + " (type: "
+              + metric.getClass()
+              + ")");
```
Comment (medb) [inline gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GhfsInstrumentation.java:265]: Use String.format

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStream.java [hunk 4]
```diff
@@ -114,16 +123,19 @@ public synchronized int read(byte[] buf, int offset, int length) throws IOExcept
...
       throw new IndexOutOfBoundsException();
     }
-
+    streamStatistics.readOperationStarted(getPos(), length);
     int numRead = channel.read(ByteBuffer.wrap(buf, offset, length));
 
     if (numRead > 0) {
       // -1 means we actually read 0 bytes, but requested at least one byte.
       statistics.incrementBytesRead(numRead);
       statistics.incrementReadOps(1);
```
Comment (medb) [inline gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStream.java:132]: Do we still need this `statistics`?

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemBase.java [hunk 5]
```diff
@@ -107,18 +109,10 @@
...
-import org.apache.hadoop.fs.FSDataOutputStream;
-import org.apache.hadoop.fs.FileAlreadyExistsException;
-import org.apache.hadoop.fs.FileChecksum;
-import org.apache.hadoop.fs.FileStatus;
-import org.apache.hadoop.fs.FileSystem;
-import org.apache.hadoop.fs.GlobPattern;
-import org.apache.hadoop.fs.Path;
-import org.apache.hadoop.fs.PathFilter;
-import org.apache.hadoop.fs.XAttrSetFlag;
+import org.apache.hadoop.fs.*;
```
Comment (medb) [inline gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemBase.java:112]: Revert

**Review by medb — CHANGES_REQUESTED** [review]

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/AbstractGhfsStatisticsSource.java (hunk 1 above)
Comment (Nivedhita-Chandran) [reply gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/AbstractGhfsStatisticsSource.java]: Updated

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/InstrumentatedGoogleHadoopFileSystem.java [hunk 6]
```diff
@@ -37,7 +38,12 @@
 
 public class InstrumentatedGoogleHadoopFileSystem extends GoogleHadoopFileSystemBase
     implements IOStatisticsSource {
-  GhfsInstrumentation instrumentation;
+
+  /** Instrumentation to track Statistics */
+  private GhfsInstrumentation instrumentation;
```
Comment (mprashanthsagar) [inline gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/InstrumentatedGoogleHadoopFileSystem.java:43]: IF this class is only responsible for IOStatistics, can we rename it accordingly ?

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/InstrumentatedGoogleHadoopFileSystem.java [hunk 7]
```diff
@@ -295,6 +302,28 @@ protected void incrementStatistic(GhfsStatistic statistic, long count) {
...
+   * Create the storage statistics or bind to an existing one.
+   *
+   * @param ioStatistics IOStatistics to build the storage statistics from.
+   * @return a storage statistics instance; expected to be that of the FS.
+   */
+  protected static GhfsStorageStatistics createStorageStatistics(final IOStatistics ioStatistics) {
+    return (GhfsStorageStatistics)
+        GlobalStorageStatistics.INSTANCE.put(
+            GhfsStorageStatistics.NAME, () -> new GhfsStorageStatistics(ioStatistics));
+  }
```
Comment (mprashanthsagar) [inline gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/InstrumentatedGoogleHadoopFileSystem.java:315]: Let's avoid a 1-line method which has no re-usability

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java [hunk 8]
```diff
@@ -264,6 +264,28 @@ public void open_throwsExceptionWhenHadoopPathNull() {
...
+    fout.writeBytes("data");
+    fout.close();
+    myGhfs.open(new Path("/directory1/file1"));
+    GhfsStorageStatistics ss1 = myGhfs.getStorageStatistics();
+    System.out.println("STORAGE STATS ***");
+    System.out.println(ss1.getLong("op_create"));
+    System.out.println(ss1.isTracked("op_list_located_status"));
+    System.out.println(ss1.getLong("op_list_located_status"));
+    System.out.println(ss1.isTracked("op_open"));
+    System.out.println(ss1.getLong("op_open"));
```
Comment (mprashanthsagar) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java:285]: please avoid sysout

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java [hunk 9]
```diff
@@ -264,6 +264,28 @@ public void open_throwsExceptionWhenHadoopPathNull() {
...
+    assertThat(myGhfs.delete(testRoot, /* recursive= */ true)).isTrue();
```
Comment (mprashanthsagar) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java:286]: Please add assertions for storage statistics

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GhfsStorageStatistics.java [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+   * @param ioStatistics
+   */
+  public GhfsStorageStatistics(final IOStatistics ioStatistics) {
+    super(NAME, "Ghfs", ioStatistics);
+  }
+
+  /** Create the instance of Empty storage Statistics */
+  public GhfsStorageStatistics() {
+    super(NAME, "Ghfs", emptyStatistics());
+  }
```
Comment (mprashanthsagar) [inline gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GhfsStorageStatistics.java:45]: Do we need this constructor ?

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java [hunk 11]
```diff
@@ -1394,4 +1596,18 @@ public void unauthenticatedAccessToPublicBuckets_googleCloudProperties() throws
 
     assertThat(fileStatuses).isNotEmpty();
   }
+
+  private static void initializeInMemoryFileSystem(FileSystem ghfs, String initUriString)
```
Comment (mprashanthsagar) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java:1600]: I dont see usages of this method, can we remove it ?

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/HadoopFileSystemTestBase.java [hunk 12]
```diff
@@ -275,6 +275,73 @@ protected void validateListFileInfo(
...
+
+    // Write an object.
+    URI path = GoogleCloudStorageFileSystemIntegrationTest.getTempFilePath();
+    Path hadoopPath = ghfsHelper.castAsHadoopPath(path);
+    String text = "Hello World!";
+
+    int numBytesWritten = ghfsHelper.writeFile(hadoopPath, text, 1, /* overwrite= */ false);
+    GhfsStorageStatistics StorageStats = ((GoogleHadoopFileSystem) ghfs).getStorageStatistics();
+
+    // Check the statistics related to seek() method.
```
Comment (mprashanthsagar) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/HadoopFileSystemTestBase.java:322]: please remove duplicate comment

**Review by mprashanthsagar — CHANGES_REQUESTED** [review]
Can we add some integration tests to cover both iostatistics and storage statistics ?

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/HadoopFileSystemTestBase.java (hunk 12 above)
Comment (Nivedhita-Chandran) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/HadoopFileSystemTestBase.java:322]: Removed

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java (hunk 11 above)
Comment (Nivedhita-Chandran) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java]: Removed the method

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GhfsStorageStatistics.java (hunk 10 above)
Comment (Nivedhita-Chandran) [reply gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GhfsStorageStatistics.java]: No. Removed the constructor

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java (hunk 9 above)
Comment (Nivedhita-Chandran) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java:281]: Added

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java (hunk 8 above)
Comment (Nivedhita-Chandran) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java]: Updated

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/InstrumentatedGoogleHadoopFileSystem.java (hunk 7 above)
Comment (Nivedhita-Chandran) [reply gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/InstrumentatedGoogleHadoopFileSystem.java]: Updated

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/InstrumentatedGoogleHadoopFileSystem.java (hunk 6 above)
Comment (Nivedhita-Chandran) [reply gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/InstrumentatedGoogleHadoopFileSystem.java:42]: We are using this class for also Storage Statistics

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemBase.java [hunk 13]
```diff
@@ -152,7 +153,7 @@
  * throw or to return false.
  */
 public abstract class GoogleHadoopFileSystemBase extends FileSystem
-    implements FileSystemDescriptor {
+    implements FileSystemDescriptor, IOStatisticsSource {
```
Comment (mprashanthsagar) [inline gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemBase.java:156]: Do we need this ? 
Also please remove all the cosmetic changes in the file

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GhfsStorageStatistics.java [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+public class GhfsStorageStatistics extends StorageStatisticsFromIOStatistics {
+
+  public static final String NAME = "GhfsStorageStatistics";
+
+  /**
+   * Create the Storage Statistics instance from the IOStatistics
+   *
+   * @param ioStatistics
+   */
+  public GhfsStorageStatistics(final IOStatistics ioStatistics) {
```
Comment (mprashanthsagar) [inline gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GhfsStorageStatistics.java:36]: Can we remove the final qualifier for the argument ?

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/InstrumentatedGoogleHadoopFileSystem.java [hunk 15]
```diff
@@ -294,6 +303,29 @@ protected void incrementStatistic(GhfsStatistic statistic) {
...
+  //   *
+  //   * @param ioStatistics IOStatistics to build the storage statistics from.
+  //   * @return a storage statistics instance; expected to be that of the FS.
+  //   */
+  //  protected static GhfsStorageStatistics createStorageStatistics(final IOStatistics
+  // ioStatistics) {
+  //    return (GhfsStorageStatistics)
+  //        GlobalStorageStatistics.INSTANCE.put(
+  //            GhfsStorageStatistics.NAME, () -> new GhfsStorageStatistics(ioStatistics));
+  //  }
```
Comment (mprashanthsagar) [inline gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/InstrumentatedGoogleHadoopFileSystem.java:318]: Please remove commented code

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java [hunk 16]
```diff
@@ -264,6 +264,23 @@ public void open_throwsExceptionWhenHadoopPathNull() {
...
+    Path testRoot = new Path("/directory1/");
+    myGhfs.mkdirs(testRoot);
+
+    FSDataOutputStream fout = myGhfs.create(new Path("/directory1/file1"));
+
+    fout.writeBytes("data");
+    fout.close();
+    myGhfs.open(new Path("/directory1/file1"));
+    GhfsStorageStatistics storage_stats = myGhfs.getStorageStatistics();
+    assertThat(storage_stats.isTracked("op_open")).isTrue();
```
Comment (mprashanthsagar) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java:280]: Are there any other metrics we want assert here ?

**Review by mprashanthsagar — CHANGES_REQUESTED** [review]

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemBase.java (hunk 13 above)
Comment (Nivedhita-Chandran) [reply gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemBase.java]: Removed

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GhfsStorageStatistics.java (hunk 14 above)
Comment (Nivedhita-Chandran) [reply gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GhfsStorageStatistics.java]: Removed

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/InstrumentatedGoogleHadoopFileSystem.java (hunk 15 above)
Comment (Nivedhita-Chandran) [reply gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/InstrumentatedGoogleHadoopFileSystem.java]: Removed

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java (hunk 16 above)
Comment (Nivedhita-Chandran) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java:280]: This method is to validate only open related statistics. Other metrics are validated in different methods

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemBase.java [hunk 17]
```diff
@@ -122,6 +122,7 @@
 import org.apache.hadoop.fs.impl.AbstractFSBuilderImpl;
 import org.apache.hadoop.fs.impl.OpenFileParameters;
 import org.apache.hadoop.fs.permission.FsPermission;
+import org.apache.hadoop.fs.statistics.IOStatisticsSource;
```
Comment (mprashanthsagar) [inline gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemBase.java:125]: Please revert all the changes in the file, they are no-op

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java [hunk 18]
```diff
@@ -702,6 +719,181 @@ public void testConfigureBucketsWithRootBucketButNoSystemBucket() throws IOExcep
...
+    GoogleHadoopFileSystem myGhfs = createInMemoryGoogleHadoopFileSystem();
+    URI fileUri = GoogleCloudStorageFileSystemIntegrationTest.getTempFilePath();
+    Path filePath = ghfsHelper.castAsHadoopPath(fileUri);
+    GhfsStorageStatistics StorageStats = myGhfs.getStorageStatistics();
+
+    try (FSDataOutputStream fout = myGhfs.create(filePath, true, 1, (short) 1, 1)) {
+      fout.writeBytes("Test Content");
+      fout.close();
+    }
+    assertThat(StorageStats.isTracked("op_create")).isTrue();
```
Comment (mprashanthsagar) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java:733]: Storagestats also implements `getLong(String key)` . Could you please assert this method as well. To ensure correctness of the values ?  ** for al tests **

**Review by mprashanthsagar — CHANGES_REQUESTED** [review]

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemBase.java (hunk 17 above)
Comment (Nivedhita-Chandran) [reply gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemBase.java]: Reverted the changes

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java (hunk 18 above)
Comment (Nivedhita-Chandran) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemIntegrationTest.java:717]: Updated with tests for getLong(String key)

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/HadoopFileSystemTestBase.java [hunk 19]
```diff
@@ -275,6 +275,72 @@ protected void validateListFileInfo(
...
+
+      // Check the statistics of method read(buf, off, len)
+      byte[] readbuffer = new byte[1];
+      readStream.read(readbuffer, 0, 1);
+      assertThat(StorageStats.isTracked("stream_read_operations")).isTrue();
+      assertThat(StorageStats.isTracked("stream_read_bytes")).isTrue();
+
+      // Check the statistics of method read(pos, buf, off, len)
+      byte[] readbuffer1 = new byte[2];
+      readStream.read(2, readbuffer1, 0, 2);
```
Comment (mprashanthsagar) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/HadoopFileSystemTestBase.java:302]: please assert values

Comment (Nivedhita-Chandran) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/HadoopFileSystemTestBase.java:303]: Updated

Comment (mprashanthsagar) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/HadoopFileSystemTestBase.java:303]: Shouldn't the value increase with new calls ?

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/HadoopFileSystemTestBase.java [hunk 20]
```diff
@@ -275,6 +275,76 @@ protected void validateListFileInfo(
...
+    String text = "Hello World!";
+    int numBytesWritten = ghfsHelper.writeFile(hadoopPath, text, 1, /* overwrite= */ false);
+    GhfsStorageStatistics StorageStats =
+        new GhfsStorageStatistics(((GoogleHadoopFileSystem) ghfs).getIOStatistics());
+    try (FSDataInputStream readStream = ghfs.open(hadoopPath)) {
+
+      //  Check the statistics of method read().
+      readStream.read();
+      assertThat(StorageStats.isTracked("stream_read_operations")).isTrue();
+      assertThat(StorageStats.getLong("stream_read_operations")).isEqualTo(2);
```
Comment (mprashanthsagar) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/HadoopFileSystemTestBase.java:293]: why do we have two calls after first read ?

Comment (Nivedhita-Chandran) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/HadoopFileSystemTestBase.java]: Updated

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/HadoopFileSystemTestBase.java [hunk 21]
```diff
@@ -275,6 +275,73 @@ protected void validateListFileInfo(
...
+    }
+    try (FSDataInputStream readStream = ghfs.open(hadoopPath)) {
+      GhfsStorageStatistics StorageStats =
+          new GhfsStorageStatistics((readStream).getIOStatistics());
+      // Check the statistics of method read(pos, buf, off, len)
+      byte[] readbuffer = new byte[2];
+      readStream.read(2, readbuffer, 0, 2);
+      readStream.close();
+      assertThat(StorageStats.isTracked("stream_read_operations")).isTrue();
+      assertThat(StorageStats.getLong("stream_read_operations")).isEqualTo(2);
```
Comment (mprashanthsagar) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/HadoopFileSystemTestBase.java:313]: shouldn't this be a single operation ?

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemTestBase.java [hunk 22]
```diff
@@ -447,6 +447,31 @@ public void testMakeQualifiedRoot() {
...
+        new GhfsStorageStatistics(((GoogleHadoopFileSystem) ghfs).getIOStatistics());
+
+    // Temporary file in local FS.
+    File localTempFile = File.createTempFile("ghfs-test-", null);
+    Path localTempFilePath = new Path(localTempFile.getPath());
+
+    ghfs.copyFromLocalFile(false, true, localTempFilePath, tempDirPath);
+
+    assertThat((StorageStats.isTracked("op_copy_from_local_file"))).isTrue();
+    assertThat((StorageStats.getLong("op_copy_from_local_file"))).isEqualTo(3);
```
Comment (mprashanthsagar) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemTestBase.java:468]: Isn't this a single request to copy ?

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/HadoopFileSystemTestBase.java (hunk 21 above)
Comment (Nivedhita-Chandran) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/HadoopFileSystemTestBase.java:313]: Here the length is given as 2. So the no of read operation called will be 2.

Comment (mprashanthsagar) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/HadoopFileSystemTestBase.java:313]: As discussed offline, The metric indicates number of calls to the API, and we are double counting in our impl.

Comment (Nivedhita-Chandran) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/HadoopFileSystemTestBase.java]: Updated

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemTestBase.java (hunk 22 above)
Comment (Nivedhita-Chandran) [reply gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemTestBase.java]: Updated

**Review by mprashanthsagar — APPROVED** [review]

**Pull request merged**